File: src/cs/TreeCensus/TreeSpecies.cs
using System;

namespace TreeCensus
{
    /// <summary>
    /// Represents a tree species.
    /// </summary>
    public class TreeSpecies : IEquatable<TreeSpecies>, ISpecies
    {
        /// <inheritdoc/>
        public string LatinName { get; }

        /// <inheritdoc/>
        public string CommonName { get; }

        /// <summary>
        /// Initializes a new instance of the <see cref="TreeSpecies"/> class.
        /// </summary>
        /// <param name="commonName">The common name.</param>
        /// <param name="latinName">The Latin name.</param>
        /// <exception cref="ArgumentNullException"><paramref name="commonName"/> is <see langword="null"/>.</exception>
        /// <exception cref="ArgumentNullException"><paramref name="latinName"/> is <see langword="null"/>.</exception>
        public TreeSpecies(string commonName, string latinName)
        {
            if (commonName == null)
            {
                throw new ArgumentNullException(nameof(commonName));
            }
            else if (latinName == null)
            {
                throw new ArgumentNullException(nameof(latinName));
            }
            else
            {
                CommonName = commonName;
                LatinName = latinName;
            }
        }

        /// <inheritdoc/>
        public override bool Equals(object? obj)
        {
            return Equals(obj as TreeSpecies);
        }

        /// <inheritdoc/>
        public bool Equals(TreeSpecies? other)
        {
            return other != null && CommonName.Equals(other.CommonName, StringComparison.OrdinalIgnoreCase) && LatinName.Equals(other.LatinName, StringComparison.OrdinalIgnoreCase);
        }

        /// <inheritdoc/>
        public override int GetHashCode()
        {
            HashCode result = new HashCode();

            result.Add(CommonName);
            result.Add(LatinName);

            return result.ToHashCode();
        }
    }
}

File: src/project2/TreeSpecies.java
package project2;

import java.util.Objects;

/**
 * Represents a single tree species with a common and Latin name.
 */
public class TreeSpecies {
    private String latinName;
    private String commonName;

    /**
     * Initializes a new instance of the {@link TreeSpecies} class.
     * 
     * @param commonName the common name.
     * @param latinName  the scientific name.
     */
    public TreeSpecies(String commonName, String latinName) {
        if (commonName == null) {
            throw new IllegalArgumentException("Value cannot be null. Argument name: commonName.");
        } else if (latinName == null) {
            throw new IllegalArgumentException("Value cannot be null. Argument name: latinName.");
        } else {
            this.commonName = commonName;
            this.latinName = latinName;
        }
    }

    /**
     * Gets the scientific (Latin) name of the tree's species. This property
     * represents the {@code spc_latin} field in the dataset.
     * 
     * @return The scientific name, a non-null string.
     */
    public String getLatinName() {
        return latinName;
    }

    /**
     * Gets the common (English) name of the tree's species. This property
     * represents the {@code spc_common} field in the dataset.
     * 
     * @return The common name, a non-null string.
     */
    public String getCommonName() {
        return commonName;
    }

    /** {@inheritDoc} */
    @Override
    public boolean equals(Object obj) {
        if (obj == null || !(obj instanceof TreeSpecies)) {
            return false;
        } else {
            final TreeSpecies other = (TreeSpecies) obj;

            return commonName.equalsIgnoreCase(other.commonName) && latinName.equalsIgnoreCase(other.latinName);
        }
    }

    /** {@inheritDoc} */
    @Override
    public int hashCode() {
        // Hash upper-case strings to preserve case-insensitivity

        return Objects.hash(latinName.toUpperCase(), commonName.toUpperCase());
    }
}
